### PR TITLE
BFD-2220: ccw field additions and updates to the clm_drg_code

### PR DESCRIPTION
**Pull request by cbrunefearless**
**JIRA Ticket:**
[BFD-2220](https://jira.cms.gov/browse/BFD-2220)

**User Story or Bug Summary:**

User Story: As a BFD user, I want BFD to seamlessly handle CCW's DRG code expansion from 3 characters to 4 characters so that my data pulls from BFD are unaffected by this change.

User Story: As a BFD Engineer, I want to make the necessary database schema change for DRG_CD (3->4 characters), CLAIM_QUERY_CODE,  and CARR_CLM_BLG_NPI_NUM so that BFD users are not affected by this change.

 

AC:

Complete database migration scripts for the following:
Updating the DRG_CD (3 -> 4 characters)
Adding the CLAIM_QUERY_CODE (Assumption the query code is already on some other claim types so it will be the same data type and one the same tables as the current one. CLAIM_QUERY_CODE: We already receive this for Inpatient, Outpatient and SNF claim types. We are requesting it for HHA and Hospice. See CLAIM_QUERY_CODE on the Inpatient tab of the RIF XLS Inpatient tab. In the BFD db, the column is called CLAIM_QUERY_CODE, with type = CHAR(1) and resides in the claim level tables e.g. INPATIENT_CLAIMS.)
Adding the CARR_CLM_BLG_NPI_NUM(Assumption that the NPI will be the same datatype as other NPIs and on the same tables as the current one. This is for Carrier claims only and will be stored at the claim level e.g. CARRIER_CLAIMS as it will ultimately populate Eob.provider. This is similar to our current ingestion of PRVDR_NUM and/or ORG_NPI_NUM. Likely a CHAR(10) field)
Test on a clone of the database to ensure that there is no impact to the API (i.e. long running migration that locks out reads)
New claim field will default to null for existing data and will be populated with null for new records
 

Open Question:

Discuss requirements for testing using the FHIR server
 

Implementation Notes:

Spreadsheet to be updated so that new fields are available via getter/setter
Will need special casing for the column enums to exclude from the RIF header
 

General Notes:

 

Confirmed that 3 character code will continue to be valid
Can go in before 12/2
Required for CCW Batch Target 1 go-live
Needs testing on a clone to understand locking implications if any
If goes in before 12/2 need to handle the new record case (null setting)
 

---

### What Does This PR Do?

This pr is for database schema changes to update the drg_code from 3 to 4 chracters and add two new fields.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [] Yes
      * [X] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] The data dictionary has been updated with any field mapping changes, if any were made.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V97__update_drgcode_add_claimquerycode_and_carrclmblgnpi.sql
Comment (mackec1) [inline apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V97__update_drgcode_add_claimquerycode_and_carrclmblgnpi.sql:6]: I'd get rid of the ${logic-psql-only} throughout the script. You only need to do that if there is a difference between HSQL and PSQL commands which I don't believe there are in this case.

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V97__update_drgcode_add_claimquerycode_and_carrclmblgnpi.sql
Comment (mackec1) [inline apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V97__update_drgcode_add_claimquerycode_and_carrclmblgnpi.sql:7]: Also, don't need theCOLLATE pg_catalog."default"

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V97__update_drgcode_add_claimquerycode_and_carrclmblgnpi.sql
Comment (mackec1) [inline apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V97__update_drgcode_add_claimquerycode_and_carrclmblgnpi.sql:11]: I can't find claim_query_code in the HHA or HOSPICE sheet(s); and can't find carr_clm_blg_npi_num in Carrier sheet.

**Review by mackec1 — COMMENTED** [review]
left a couple of comments.

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V97__update_drgcode_add_claimquerycode_and_carrclmblgnpi.sql
Comment (cbrunefearless) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V97__update_drgcode_add_claimquerycode_and_carrclmblgnpi.sql]: done

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V97__update_drgcode_add_claimquerycode_and_carrclmblgnpi.sql
Comment (cbrunefearless) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V97__update_drgcode_add_claimquerycode_and_carrclmblgnpi.sql]: done

**Review by mackec1 — DISMISSED** [review]
Thx for clarifying the XSLX; forgot this was discussed in sprint planning.

**Review by mackec1 — APPROVED** [review]

**Review by mackec1 — APPROVED** [review]

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V97__update_drgcode_add_claimquerycode_and_carrclmblgnpi.sql
Comment (christrotteradhoc) [inline apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V97__update_drgcode_add_claimquerycode_and_carrclmblgnpi.sql:14]: Making this comment here since I cannot comment on the spreadsheet:

Since these were added to the spreadsheet, I believe they will become columns on the generated entities, right? How are they being suppressed from the columns expected to be in the RIF?

Also, for DRG_CD, we need to have tests for both 3 and 4 character DRG_CD values in the RIFs

Comment (cbrunefearless) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V97__update_drgcode_add_claimquerycode_and_carrclmblgnpi.sql:14]: Claim query code and carrier claim blg npi were not added to the spreadsheet.. The only change was to the  clm_drg_cd in inpatient and snf to change it from 3 to 4 characters.. Will add tests to the snf and inpatient entities.

Comment (cbrunefearless) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V97__update_drgcode_add_claimquerycode_and_carrclmblgnpi.sql:14]: Added temporary code to not include those two new fields

**Review by christrotteradhoc — COMMENTED** [review]
Pulled down the spreadsheet and built the code and things look as I expect. Thanks for getting those changes in too. Would like to hold this change on a branch until we get confirmation from CCW and an updated requirements doc to ensure names, types, and length match as we expect they will.

**Review by mackec1 — APPROVED** [review]
Looks reasonable.

**Review by mackec1 — APPROVED** [review]
LGTM

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayout.java
Comment (lsmitchell) [inline apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayout.java:134]: Are these going away before the merge to master? The when should be described here so we don't miss it. (If its further than "when we merge this to master eventually" it should also probably have a ticket and be linked here)

File: apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/test/java/gov/cms/bfd/pipeline/ccw/rif/extract/RifFilesProcessorTest.java
Comment (lsmitchell) [inline apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/test/java/gov/cms/bfd/pipeline/ccw/rif/extract/RifFilesProcessorTest.java:857]: Maybe throw a comment here that this is the field we're primarily testing, just cause it's hard to find in the noise

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V98__update_drgcode_add_claimquerycode_and_carrclmblgnpi.sql
Comment (lsmitchell) [inline apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V98__update_drgcode_add_claimquerycode_and_carrclmblgnpi.sql:14]: Are these not tested here because other stories are adding tests/impl for them?

Comment (cbrunefearless) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V98__update_drgcode_add_claimquerycode_and_carrclmblgnpi.sql:14]: yes they are in PR-2221 and PR-2222

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/InpatientClaimTransformerV2Test.java
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/InpatientClaimTransformerV2Test.java:381]: done

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/SNFClaimTransformerV2Test.java
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/SNFClaimTransformerV2Test.java:411]: done

**Review by lsmitchell — APPROVED** [review]

**Pull request merged**
